### PR TITLE
Dragon drop

### PR DESCRIPTION
**Pull request by gtodd876**

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -303,3 +308,36 @@ function appendStory(story, board, storyUrl, statusBadge, estimate) {
...
+      updatePivotal(ui);
+  });
+} //end initializeDragAndDrop
+function updatePivotal(ui) {
+    let id = ui.item.data("id");
+    console.log(ui.item);
+    let newStatus = ui.item.closest(".js-board").data("status");
+    let beforeId = ui.item.prev().data("id");
+    let urlId = "/" + ui.item.data("id");
+    if(beforeId === undefined) beforeId = null;
```
Comment (coxandrew) [inline script.js]: missing space before conditional. Should be:

```
if (beforeId === undefined) beforeId = null;
```

File: script.js [hunk 2]
```diff
@@ -303,3 +308,36 @@ function appendStory(story, board, storyUrl, statusBadge, estimate) {
...
+
+function initializeDragAndDrop () {
+  $(".todo, .in-progress, .ready-for-review, .done").sortable({
+    connectWith: ".todo, .in-progress, .ready-for-review, .done"
+  });
+
+  $(".js-board").on("sortstop", function(event, ui) {
+      updatePivotal(ui);
+  });
+} //end initializeDragAndDrop
```
Comment (coxandrew) [inline script.js]: You should be able to simplify this with (tested locally):

```
function initializeDragAndDrop () {
  $(".js-board").sortable({
    connectWith: ".js-board",
    stop: function(event, ui) {
      updatePivotal(ui);
    }
  });
} //end initializeDragAndDrop
```

File: script.js [hunk 3]
```diff
@@ -303,3 +308,36 @@ function appendStory(story, board, storyUrl, statusBadge, estimate) {
...
+function updatePivotal(ui) {
+    let id = ui.item.data("id");
+    console.log(ui.item);
+    let newStatus = ui.item.closest(".js-board").data("status");
+    let beforeId = ui.item.prev().data("id");
+    let urlId = "/" + ui.item.data("id");
+    if(beforeId === undefined) beforeId = null;
+    instance.put("/" + id, {
+    "current_state": newStatus,
+})
```
Comment (coxandrew) [inline script.js]: The sort order within a list is not working. Looks like you're not passing the `before_id` at all? Or `after_id` if it's at the end of the list?

**Review by coxandrew — COMMENTED** [review]
I found an issue where chores can't be moved to the "finished" state. I would ignore that for now, but create a new story for handling that error case. You basically want to put the card back to where it was if someone tries to drag it there (chores don't need to be "approved" in Pivotal Tracker).

File: script.js (hunk 1 above)
Comment (gtodd876) [reply script.js]: edited

File: script.js (hunk 3 above)
Comment (gtodd876) [reply script.js]: opps! i only did it in my mind!

File: script.js (hunk 2 above)
Comment (gtodd876) [reply script.js]: edited

File: script.js (hunk 3 above)
Comment (coxandrew) [reply script.js]: What happens if it's first in the list?

File: script.js [hunk 4]
```diff
@@ -303,3 +311,36 @@ function appendStory(story, board, storyUrl, statusBadge, estimate) {
...
+function updatePivotal(ui) {
+    let id = ui.item.data("id");
+    console.log(ui.item);
+    let newStatus = ui.item.closest(".js-board").data("status");
+    let beforeId = ui.item.prev().data("id");
+    let urlId = "/" + ui.item.data("id");
+    if (beforeId === undefined) beforeId = null;
+    instance.put("/" + id, {
+    "current_state": newStatus,
+    "before_id": beforeId
```
Comment (coxandrew) [inline script.js]: Not sure where my comment went, but what happens if you move the card to the first position in the list and there's nothing before it?

Comment (gtodd876) [reply script.js]: i dont get any errors and they look like they are in order on my app

File: script.js [hunk 5]
```diff
@@ -68,29 +72,33 @@ function displayStories(
...
+    appendStory(story, board, storyUrl, statusBadge, estimate, storyId);
   });
   doneStories.forEach(function(story) {
     let board = "done";
     let storyUrl = story.url;
     let statusBadge = story.current_state;
     let estimate = story.estimate;    
-    appendStory(story, board, storyUrl, statusBadge, estimate);
+    let storyId = story.id; 
+    appendStory(story, board, storyUrl, statusBadge, estimate, storyId);
```
Comment (coxandrew) [inline script.js]: There's a lot of duplication in these setup loops. Is there any way you can DRY it up?

File: script.js [hunk 6]
```diff
@@ -303,3 +311,53 @@ function appendStory(story, board, storyUrl, statusBadge, estimate) {
...
+function updatePivotal(ui) {
+    let id = ui.item.data("id");
+    let newStatus = ui.item.closest(".js-board").data("status");
+    let afterId = ui.item.prev().data("id");
+    let beforeId = ui.item.next().data("id");
+    let urlId = "/" + ui.item.data("id");
+    let currentCard = ui.item;
+    if (beforeId === undefined) beforeId = null;
+    if (afterId === undefined) afterId = null;
+    if (newStatus === 'accepted') {
```
Comment (coxandrew) [inline script.js]: Use double quotes for consistency.

File: script.js [hunk 7]
```diff
@@ -303,3 +311,53 @@ function appendStory(story, board, storyUrl, statusBadge, estimate) {
...
+        
+      })
+      .catch(function(error) {
+      console.log(error);
+      });
+    } else {
+      instance.put("/" + id, {
+        "current_state": newStatus,
+        "after_id": afterId,
+        "before_id": beforeId
```
Comment (coxandrew) [inline script.js:334]: If both `after_id` and `before_id` are `nil`, would it work to pass them along for an accepted story? Then we might be able to get rid of this duplicated branch code.

Comment (gtodd876) [reply script.js:334]: Even if they are null, if i pass `before_id` and `after_id`  i get PUT error from Pivotal :(

File: script.js [hunk 8]
```diff
@@ -1,8 +1,10 @@
+let mykey = config.MY_API_KEY;
+
 let instance = axios.create({
   baseURL:
-    "https://www.pivotaltracker.com/services/v5/projects/2111883/stories/",
-  timeout: 1000,
-  headers: { "X-TrackerToken": "849f56353d70242c0a95548fc4b022cd" }
+  "https://www.pivotaltracker.com/services/v5/projects/2111883/stories",
```
Comment (coxandrew) [inline script.js]: You should have a `project_id` in your `config.js` as well.

File: script.js [hunk 9]
```diff
@@ -288,13 +266,15 @@ function getEmailAddress(id) {
   } else return "test@test.com";
 }
 
-function appendStory(story, board, storyUrl, statusBadge, estimate) {
+function appendStory(story, storyUrl, statusBadge, estimate, storyId) {
   let storyType = story.story_type;
   let email = getEmailAddress(story.owned_by_id);
+  //console.log(story, storyUrl, statusBadge, estimate, storyId, email, storyType);
```
Comment (coxandrew) [inline script.js]: Remove commented out log statement.

File: script.js [hunk 10]
```diff
@@ -288,13 +266,15 @@ function getEmailAddress(id) {
...
   let urlForGravatar = gravatar(email, {size: "30"});
   if (estimate === undefined) estimate = "";
-  $("." + board).append(
-    "<div class='card " + storyType + "'>" +
+  // console.log($("div[data-accepts='started']").get());
```
Comment (coxandrew) [inline script.js]: Log statement.

File: script.js [hunk 11]
```diff
@@ -288,13 +266,15 @@ function getEmailAddress(id) {
...
+  $("div[data-accepts~='" + statusBadge + "']").append(
+    "<div class='card " + storyType + "' data-id='" + storyId + "'>" +
```
Comment (coxandrew) [inline script.js]: I dig it!

File: script.js [hunk 12]
```diff
@@ -20,79 +22,55 @@ instance
...
+        allStories[i].current_state === "accepted"
+      )
+      doneStories.push(allStories[i]);
     }
-
-    if (allStories[i].current_state === "finished") {
-      reviewStories.push(allStories[i]);
+  }
+  
+  function displayStories(todoStories, reviewStories, inprogressStories, doneStories) {
```
Comment (coxandrew) [inline script.js]: I believe you can use the `rest args` feature here (might need to check browser compatibility):

```
function displayStories(...storyColumns) {
```

Then you can just call `storyColumns.forEach()` in the function.

**Review by coxandrew — CHANGES_REQUESTED** [review]
When you're done and ready for the review, remove the `[WIP]`.

**Review by coxandrew — APPROVED** [review]
LGTM!

**Pull request merged**